[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38447 [Tensorpipe Agent] Enabling tests with OSS CI
* #38446 [Tensorpipe Agent] Dist Optimizer Tests for Tensorpipe Agent
* #38445 [Tensorpipe Agent] Dist Autograd Tests for Tensorpipe Agent
* **#38444 [Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent**

This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

osalpekar committed:
[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent

This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -304,6 +307,15 @@ def get_rref_debug_info():
 
 class RpcTest(RpcAgentTestFixture):
+    def _skip_if_tensorpipe_agent(old_func):  # noqa
+        def decorator(self):
+            return unittest.skipIf(
+                self.rpc_backend == rpc.backend_registry.BackendType.TENSORPIPE,
+                "This test is not yet supported in the Tensorpipe Agent"
+            )(old_func)
```
Tip: `functools.wraps` is a nice tool to help make wrapper functions: it makes sure to also propagate some metadata from the wrapped function to the wrapper one (like docstrings, annotations, ...). I don't know if it matter here, but if you want to have a look: https://docs.python.org/3.8/library/functools.html#functools.wraps

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -516,6 +528,7 @@ def test_register_rpc_backend_and_set_and_start_rpc_backend(
         )
 
     @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    @_skip_if_tensorpipe_agent
```
There is already a decorators that seems to hint that this should only be run for the processgroup agent. Is that decorator not working? Why do we need to add another one?

In principle I think it's fine to introduce a new different decorator (`_skip_if_tensorpipe_agent`) to mark the tests that we should support but currently aren't: it makes it clear that it's a temporary measure rather than the intended situation. However, if this specific test is never supposed to run with anything else than processgroup, then we should not use this new decorator but rather make sure the old one (`requires_process_group_agent`) works.

This comment applies several more times below.

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1065,6 +1078,7 @@ def test_py_tensors(self):
         self.assertEqual(ret, my_tensor_function(torch.ones(n, n), torch.ones(n, n)))
 
     @dist_init
+    @_skip_if_tensorpipe_agent
```
In the previous test the `_skip_if_tensorpipe_agent` came before the `dist_init` decorator, here it comes after. Does it make any difference? You mentioned in the PR's summary that `dist_init` was setting some value to processgroup as default, perhaps it's due to this?

lw reviewed (approved):
I'd be surprised if the CI failures come from this diff, since if I understand this correctly you're not enrolling these tests to be run by the CI until the last PR in the stack, right?

osalpekar commented:
> I'd be surprised if the CI failures come from this diff, since if I understand this correctly you're not enrolling these tests to be run by the CI until the last PR in the stack, right?

Right, rebasing should probably fix these failures. We synced offline regarding the decorator comments, and adding functionality to Tensorpipe Agent should allow us to remove most decorators + a future test refactor to improve the process of selecting/registering backends for the test suite should streamline skipping tests based on the agent

osalpekar committed:
Update on "[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent"


This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent"


This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent"


This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -319,6 +322,15 @@ def callback(fut):
 
 
 class RpcTest(RpcAgentTestFixture):
+    def _skip_if_tensorpipe_agent(old_func):  # noqa
```
nit: do we have a type for this noqa? e.g. F401, etc.

osalpekar replied:
Yes, it's B902 - just added the types to the noqa

mrshenli commented on torch/testing/_internal/distributed/rpc/tensorpipe_rpc_agent_test_fixture.py:
```diff
@@ -0,0 +1,32 @@
+import torch.testing._internal.dist_utils as dist_utils
+
+# In order to run the existing test RPC and Distributed Autograd test suites
+# with Tensorpipe, we make introduce a new class in both rpc_test.py and
```
nit: make introduce -> introduce

mrshenli reviewed (approved):
Thanks!

osalpekar committed:
Update on "[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent"


This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent"


This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent"


This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent"


This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] RPC, RRef tests for Tensorpipe Agent"


This enables the RPC/RRef test suites to run with the Tensorpipe RPC Agent. This creates a new fixture to ensure the backend/options used are Tensorpipe, as well as a decorator to skip tests that Tensorpipe currently cannot support due to missing functionality.

One small note: the decorator function is a class method of the test class so we can check whether `self.rpc_backend` is tensorpipe. In the class-scope, the `TEST_CONFIG.rpc_backend_name` string is set to Tensorpipe, but outside the class scope, it is PGA, possibly due to importing dist_utils which sets this config to PGA by default. The cleanest solution would be to refactor the backend selection to be more uniform (since currently every backend is set slightly differently), but that would be a longer-term fix.

Differential Revision: [D21412016](https://our.internmc.facebook.com/intern/diff/D21412016/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21412016/)!

[ghstack-poisoned]
